src/Util/AnnotationFilterExtractorTrait.php

```diff
@@ -0,0 +1,146 @@
+                $properties[$reflectionProperty->getName()] = $filterAnnotation->strategy;
+            }
+        }
+
+        return $properties;
+    }
+
+    /**
+     * Reads filter annotations from a ReflectionClass.
+     *
+     * @param \ReflectionClass $reflectionClass
+     * @param Reader           $reader
+     *
+     * @return array Key is the filter id. It has two values, properties and the ApiFilter instance
+     */
+    private function readFilterAnnotations(\ReflectionClass $reflectionClass, Reader $reader): array
+    {
+        $filters = [];
+
+        foreach ($this->getFilterAnnotations($reader->getClassAnnotations($reflectionClass)) as $filterAnnotation) {
+            $filterClass = $filterAnnotation->filterClass;
+            $id = $this->generateFilterId($reflectionClass, $filterClass);
+
+            if (!isset($filters[$id])) {
+                $filters[$id] = [$filterAnnotation->arguments, $filterClass];
+            }
+
+            if ($properties = $this->getFilterProperties($filterAnnotation, $reflectionClass)) {
+                $filters[$id][0]['properties'] = $properties;
+            }
+        }
+
+        foreach ($reflectionClass->getProperties() as $reflectionProperty) {
+            foreach ($this->getFilterAnnotations($reader->getPropertyAnnotations($reflectionProperty)) as $filterAnnotation) {
+                $filterClass = $filterAnnotation->filterClass;
+                $id = $this->generateFilterId($reflectionClass, $filterClass);
+
+                if (!isset($filters[$id])) {
+                    $filters[$id] = [$filterAnnotation->arguments, $filterClass];
+                }
+
+                if ($properties = $this->getFilterProperties($filterAnnotation, $reflectionClass, $reflectionProperty)) {
+                    if (isset($filters[$id][0]['properties'])) {
+                        $filters[$id][0]['properties'] = array_merge($filters[$id][0]['properties'], $properties);
+                    } else {
+                        $filters[$id][0]['properties'] = $properties;
+                    }
+                }
+            }
+        }
+
+        $parent = $reflectionClass->getParentClass();
+
+        if (false !== $parent) {
+            return array_merge($filters, $this->readFilterAnnotations($parent, $reader));
+        }
+
+        return $filters;
+    }
+
+    /**
+     * Generates a unique, per-class and per-filter identifier prefixed by `annotated_`.
+     *
+     * @param \ReflectionClass $reflectionClass the reflection class (a ResourceClass)
+     * @param string           $filterClass     the filter class
+     *
+     * @return string
+     */
+    private function generateFilterId(\ReflectionClass $reflectionClass, string $filterClass): string
+    {
+        return 'annotated_'.Inflector::tableize(str_replace('\\', '', $reflectionClass->getName()).(new \ReflectionClass($filterClass))->getShortName());
```

Comment:
Shouldn't you use FQN to avoid conflicts? Can you use `sprintf` too to make the code more readable?

Reply:
You mean FQN on the filter name? 
About `sprintf` there'd be only one substitution so I don't think it'd change much:

```
'annotated_'.Inflector::tableize(str_replace('\\', '', $reflectionClass->getName().(new \ReflectionClass($filterClass))->getName()));

sprintf('annotated_%s, Inflector::tableize(str_replace('\\', '', $reflectionClass->getName().(new \ReflectionClass($filterClass))->getName())));
```